src/DictVectors/communicators.jl

```diff
@@ -0,0 +1,306 @@
+* [`target_segment`](@ref): defaults to selecting using fastrange to pick the segment.
+
+"""
+abstract type Communicator end
+
+"""
+    is_distributed(::Communicator)
+
+Return `true` if communicator operates over MPI.
+"""
+is_distributed(::Communicator) = true
+
+"""
+    reduce_remote(c::Communicator, op, x)
+
+Perform a reduction over MPI, by using `MPI.Allreduce`.
+"""
+reduce_remote(c::Communicator, op, x) = MPI.Allreduce(x, op, mpi_comm(c))
+
+"""
+    total_num_segments(c::Communicator, n) -> Int
+
+Return the total number of segments, including the remote ones, where `n` is number of
+local segments.
+"""
+total_num_segments(c::Communicator, n) = n * mpi_size(c)
+
+"""
+    target_segment(c::Communicator, k, num_segments) -> target, is_local
+
+Return the target segment for a key if it's local key and whether it's local or not.
+"""
+function target_segment(c::Communicator, k, num_segments)
+    total_segments = num_segments * mpi_size(c)
+    result = fastrange_hash(k, total_segments) - mpi_rank(c) * num_segments
+    return result, 1 ≤ result ≤ num_segments
+end
+
+"""
+    mpi_rank(::Communicator) -> Int
+
+Return the MPI rank of the communicator.
+"""
+mpi_rank
+
+"""
+    mpi_size(::Communicator) -> Int
+
+Return the total number of MPI ranks.
+"""
+mpi_size
+
+"""
+    mpi_comm(::Communicator) -> MPI.Comm
+
+Return the `MPI.Comm` that the communicator operates on.
+"""
+mpi_comm
+
+"""
+    copy_to_local!([::Communicator,] w::WorkingMemory, t::PDVec) -> PDVec
+
+Copy pairs in `t` from all ranks and return them as (possibly) new [`PDVec`](@ref), possibly
+using the [`WorkingMemory`](@ref) as temporary storage.
+"""
+copy_to_local!
+
+"""
+    synchronize_remote!([::Communicator,] ::WorkingMemory)
+
+Copy pairs from remote ranks to the local part of the [`WorkingMemory`](@ref).
+"""
+synchronize_remote!
+
+
+"""
+    NotDistributed <: Communicator
+
+This [`Communicator`](@ref) is used when MPI is not available.
+"""
+struct NotDistributed <: Communicator end
+
+is_distributed(::NotDistributed) = false
+
+mpi_rank(::NotDistributed) = 0
+mpi_size(::NotDistributed) = 1
+
+synchronize_remote!(::NotDistributed, w) = w
+
+function copy_to_local!(::NotDistributed, w, t)
+    return t
+end
+
+reduce_remote(::NotDistributed, _, x) = x
+
+total_num_segments(::NotDistributed, n) = n
+
+target_segment(::NotDistributed, k, num_segments) = fastrange_hash(k, num_segments), true
+
+"""
+    LocalPart <: Communicator
+
+When [`localpart`](@ref) is used, the vector's [`Communicator`](@ref) is replaced with this.
+This allows iteration and local reductions.
+"""
+struct LocalPart{C} <: Communicator
+    communicator::C
+end
+
+is_distributed(::LocalPart) = false
+mpi_rank(lp::LocalPart) = mpi_rank(lp.communicator)
+mpi_size(lp::LocalPart) = mpi_size(lp.communicator)
+
+function synchronize_remote!(::LocalPart, w)
+    throw(CommunicatorError("attemted to synchronize localpart"))
+end
+
+reduce_remote(::LocalPart, _, x) = x
```

Comment:
I'm confused. Shouldn't this return a scalar?
```suggestion
reduce_remote(::LocalPart, op, x) = reduce(op, x)
```
Or maybe throw an error, as remote reduction of a `LocalPart` does not make sense?

Reply:
The purpose of this function is just to collect and merge already reduced results. I've changed its name to make that more apparent.